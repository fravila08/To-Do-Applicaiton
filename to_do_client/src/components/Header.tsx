import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import React, { useEffect, useState } from "react";
import axios from "axios";
import { ITask } from "../App";

interface HeaderProps {
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

export interface ResponseCreateTask {
  itemCreated: boolean;
  id: number;
}

export const createTask = async (taskTitle: string): Promise<ResponseCreateTask> => {
  let response = await axios.post("newtask/", {
    name: taskTitle,
  });
  return response["data"];
};

export const Header: React.FC<HeaderProps> = ({ allTasks, setAllTasks }) => {
  const [isDisabled, setIsDisabled] = useState(true);
  const [newTask, setNewTask] = useState("");

  useEffect(() => {
    let evalInput = newTask.replaceAll(" ", "");
    if (evalInput.length >= 1 && evalInput != "") {
      setIsDisabled(false);
    } else {
      setIsDisabled(true);
    }
  }, [newTask]);

  const createNewTask = async (
    str: string,
    event?: React.FormEvent<HTMLFormElement>
  ) => {
    event?.preventDefault();
    let response = await createTask(str);
    if (response.itemCreated) {
      setAllTasks([
        ...allTasks,
        { id: response.id, title: str, completed: false },
      ]);
      setNewTask("");
    }
  };

  return (
    <Container style={{ marginBottom: "1vh" }}>
      <Row>
        <Col xs={4}></Col>
        <Col xs={8} className="formHolder">
          <Form onSubmit={(e) => createNewTask(newTask, e)}>
            <Form.Control
              value={newTask}
              id="createTaskInput"
              onChange={(e) => setNewTask(e.target.value)}
            />
            <Button
              variant="success"
              id="createTaskButton"
              disabled={isDisabled}
              type="submit"
            >
              Create
            </Button>
          </Form>
        </Col>
      </Row>
    </Container>
  );
};
