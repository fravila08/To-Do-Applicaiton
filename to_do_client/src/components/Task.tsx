import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import { ITask } from "../App";
import Form from "react-bootstrap/Form";
import axios from "axios";
import done from "../assets/done.png";
import pending from "../assets/pending.png";

export interface TaskProps {
  task: ITask;
  selectedTasks: number[];
  setSelectedTasks: (selectedTasks: number[]) => void;
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

export const changeTaskStatus = async (id: number) => {
  try {
    let response = await axios.put(`task/${id}/`);
    return response.data.changed;
  } catch (err) {
    alert(err);
    return false;
  }
};

export const Task: React.FC<TaskProps> = ({
  task,
  setSelectedTasks,
  selectedTasks,
  allTasks,
  setAllTasks,
}) => {
  const changeStatus = async (clicked: boolean, taskToChange: ITask) => {
    let response = await changeTaskStatus(taskToChange["id"]);
    if (response) {
      setAllTasks(allTasks.filter((task) => task !== taskToChange));
      taskToChange.completed = clicked;
      setAllTasks([...allTasks]);
    }
  };

  return (
    <Row className="task">
      <Col>
        <Form.Check
          type="checkbox"
          id={`taskSelectedBtn${task.id}`}
          onChange={(e) =>
            e.target.checked
              ? setSelectedTasks([...selectedTasks, task.id])
              : setSelectedTasks(selectedTasks.filter((id) => id !== task.id))
          }
        />
      </Col>
      <Col id={`task${task.id}`} xs={7} className="taskTitle">
        {task.title}
      </Col>
      <Col className="checkHolder" xs={2}>
        <Form.Check
          id={`taskCheck${task.id}`}
          type="checkbox"
          name="checkbox"
          className="check"
          checked={task.completed}
          onChange={(e) => changeStatus(e.target.checked, task)}
        />
        <Form.Label for="checkbox">
          <img className="checkImg" src={task.completed ? done : pending} />
        </Form.Label>
      </Col>
    </Row>
  );
};
