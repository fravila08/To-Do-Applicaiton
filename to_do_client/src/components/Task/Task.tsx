import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'
import { useEffect, useState } from 'react'
import { Tasks } from '../../App'

interface IdSelectionProps{
    selected: number[];
    setSelected: (selected: number[]) => void;
    task: Tasks;
}

// export function addToSelection(selected:number[],id:number, checked:boolean){
//     console.log("hello")
//     let mySelected=selected
//     if(checked){
//         mySelected.push(Number(id))
//     }
//     else{
//         let newSelected:number[]=[]
//         for(let i=0; i< mySelected.length; i++){
//             if(mySelected[i]!==id){
//                 newSelected.push(mySelected[i])
//             }
//         }
//         mySelected=newSelected
//     }
//     console.log(mySelected,'s')
//     return mySelected
// }


export const Task: React.FC<IdSelectionProps>=({setSelected, selected, task})=>{
    const [mySelected, setMySelected]=useState<number[]>(selected)
    
    useEffect(()=>{
        setMySelected(selected)
    },[selected])


    return(
        <Row style={{backgroundColor:"grey", border:"black 2px solid"}}>
            <Col xs={9}>
                {task.title}
            </Col>
            <Col xs={2}>
                <Form.Check type="checkbox" label="select" 
                onChange={(e)=>{
                    if(e.target.checked){
                    setSelected([...selected, task.id])
                    }
                    else{ setSelected(selected.filter(item=> item !== task.id))}
                    }} 
                
                />
            </Col>
        </Row>
    )
}