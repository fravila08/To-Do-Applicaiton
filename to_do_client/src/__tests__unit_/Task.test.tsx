import { Task} from "../components/Task";
import { describe, expect, it} from "vitest";
import TestRenderer from 'react-test-renderer';



describe("Will take a snapshot of the current Task compenent",()=>{
    it("will create a snapshot",()=>{
        let selected:number[]=[]
        const setSelected=(v:number[])=>{selected=v}
        const header= TestRenderer.create(<Task 
            setSelected={setSelected} 
            selected={selected} 
            task={{"id":1,"Title":"Style", "Completed":false}}
          />)
        expect(header).toMatchSnapshot()
    })
})