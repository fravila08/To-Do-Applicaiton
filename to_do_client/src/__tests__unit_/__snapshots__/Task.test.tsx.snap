// Vitest Snapshot v1

exports[`Task.tsx > will create and match a snapshot 1`] = `
<div
  className="task row"
>
  <div
<<<<<<< HEAD
    className="col-8"
    style={
      {
        "display": "flex",
      }
    }
=======
    className="taskTitle col-9"
>>>>>>> starting_react
  >
    Style
  </div>
</div>
`;
